pyannote/audio/pipelines/speaker_diarization.py

```diff
@@ -585,18 +604,25 @@ def apply(
         if not return_embeddings:
             return diarization
 
+        # this can happen when we use OracleClustering
+        if centroids is None:
+            return diarization, None
+
+        # The number of centroids may be smaller than the number of speakers
+        # in the annotation. This can happen if the number of active speakers
+        # obtained from `speaker_count` for some frames is larger than the number
+        # of clusters obtained from `clustering`. In this case, we append zero embeddings
+        # for extra speakers
+        if len(diarization.labels()) > centroids.shape[0]:
+            centroids = np.pad(centroids, ((0, len(diarization.labels()) - centroids.shape[0]), (0, 0)))
```

Comment:
This part is still kind of confusing to me. 

Are there any guarantee that these extra speakers are actually at the end of the `diarization.labels()` list?

Reply:
@hbredin again, sorry for such a delay.
I have never seen this violated in practice, but I needed some time to understand whether it is guaranteed.

In fact, it is. Those extra speakers, which do not have a designated cluster but detected solely from the excessive speaker count, are added to the annotation in `SpeakerDiarizationMixin.to_diarization` method:

```
    @staticmethod
    def to_diarization(
        segmentations: SlidingWindowFeature,
        count: SlidingWindowFeature,
    ) -> SlidingWindowFeature:
        """Build diarization out of preprocessed segmentation and precomputed speaker count

        Parameters
        ----------
        segmentations : SlidingWindowFeature
            (num_chunks, num_frames, num_speakers)-shaped segmentations
        count : SlidingWindow_feature
            (num_frames, 1)-shaped speaker count

        Returns
        -------
        discrete_diarization : SlidingWindowFeature
            Discrete (0s and 1s) diarization.
        """
        activations = Inference.aggregate(
            segmentations,
            frames=count.sliding_window,
            hamming=False,
            missing=0.0,
            skip_average=True,
        )
        # shape is (num_frames, num_speakers)

        ### ----------- HERE EXTRA SPEAKERS ARE ADDED -------------- ###
        _, num_speakers = activations.data.shape
        max_speakers_per_frame = np.max(count.data)
        if num_speakers < max_speakers_per_frame:
            activations.data = np.pad(
                activations.data, ((0, 0), (0, max_speakers_per_frame - num_speakers))   
            )
       

        extent = activations.extent & count.extent
        activations = activations.crop(extent, return_data=False)
        count = count.crop(extent, return_data=False)

        sorted_speakers = np.argsort(-activations, axis=-1)
        binary = np.zeros_like(activations.data)

        for t, ((_, c), speakers) in enumerate(zip(count, sorted_speakers)):
            for i in range(c.item()):
                binary[t, speakers[i]] = 1.0

        return SlidingWindowFeature(binary, activations.sliding_window)
```

So the extra speaker dimensions are always appended to the end of the aggregated activations. Indices of those extra dimensions (which go after all the previous indices) are preserved as integer labels after the conversion to the continuous diarization, but before the labels renaming.

The centroids are indexed accordingly to the `hard_clusters` values, hence integer labels for speakers with clusters match with the indices of corresponding vectors in `centroids` array. The indices of remaining speakers are always larger, and hence match with zero vectors in the padding part of the centroid array.

Hence, no matter the order of `diarization.labels()` in the final annotation, this reordering of centroids array works equally well for speakers with clusters and speakers without:
```
  # re-order centroids so that they match
  # the order given by diarization.labels()
  inverse_mapping = {label: index for index, label in mapping.items()}
  centroids = centroids[
      [inverse_mapping[label] for label in diarization.labels()]
  ]
  ```